File: backend/app/utils/email_service.py
import logging
import smtplib
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
import resend

from app.configs.settings import Settings

logger = logging.getLogger(__name__)

def get_otp_email_template(otp: str, purpose: str = "verification") -> str:
    if purpose == "password_reset":
        title = "Reset Your Password"
        subtitle = "Use the verification code below to reset your BudgetBuddy password."
        subject_icon = "🔐"
    else:
        title = "Verify Your Email"
        subtitle = "Use the verification code below to complete your sign in to BudgetBuddy."
        subject_icon = "✨"

    return f"""
<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>{title}</title>
</head>
<body style="margin:0;padding:0;background:#0f172a;font-family:Arial,Helvetica,sans-serif;color:#e2e8f0;">

    <table width="100%" cellpadding="0" cellspacing="0" style="padding:40px 20px;">
        <tr>
            <td align="center">

                <table width="560" cellpadding="0" cellspacing="0"
                       style="background:#1e293b;border-radius:16px;padding:40px;border:1px solid #334155;box-shadow:0 10px 25px -5px rgba(0,0,0,0.5);">

                    <tr>
                        <td align="center">
                            <div style="width:48px;height:48px;background:rgba(99,102,241,0.2);border-radius:12px;display:flex;align-items:center;justify-content:center;margin-bottom:16px;font-size:24px;line-height:48px;text-align:center;">
                                {subject_icon}
                            </div>
                            <h2 style="margin:0;color:#f8fafc;font-size:24px;font-weight:700;">
                                {title}
                            </h2>

                            <p style="margin:16px 0 24px 0;color:#94a3b8;font-size:15px;line-height:24px;">
                                {subtitle}
                            </p>

                            <div style="
                                display:inline-block;
                                background:#090d16;
                                color:#818cf8;
                                font-size:32px;
                                font-weight:bold;
                                letter-spacing:10px;
                                padding:16px 36px;
                                border-radius:12px;
                                border:1px solid #4338ca;
                                margin:12px 0 24px 0;
                            ">
                                {otp}
                            </div>

                            <p style="margin:0;color:#64748b;font-size:13px;">
                                This code will expire in <strong style="color:#94a3b8;">10 minutes</strong>.
                            </p>

                            <p style="margin-top:28px;color:#475569;font-size:12px;border-top:1px solid #334155;padding-top:20px;">
                                If you did not request this code, you can safely ignore this email. Your account remains secure.
                            </p>

                        </td>
                    </tr>

                </table>

                <p style="margin-top:24px;color:#475569;font-size:12px;">
                    © 2026 BudgetBuddy. All rights reserved.
                </p>

            </td>
        </tr>
    </table>

</body>
</html>
"""

def send_smtp_email(to_email: str, subject: str, html_content: str) -> bool:
    """Send email via SMTP as fallback."""
    if not (Settings.SMTP_SERVER and Settings.SENDERS_EMAIL and Settings.GMAIL_PASSWORD):
        logger.warning("SMTP configuration is incomplete.")
        return False
    try:
        msg = MIMEMultipart()
        msg["Subject"] = subject
        msg["From"] = Settings.SENDERS_EMAIL
        msg["To"] = to_email
        msg.attach(MIMEText(html_content, "html"))

        with smtplib.SMTP(Settings.SMTP_SERVER, Settings.SMTP_PORT) as server:
            server.starttls()
            server.login(Settings.SENDERS_EMAIL, Settings.GMAIL_PASSWORD)
            server.sendmail(Settings.SENDERS_EMAIL, to_email, msg.as_string())
        logger.info(f"Email successfully sent via SMTP to {to_email}")
        return True
    except Exception as e:
        logger.error(f"Failed to send email via SMTP: {e}")
        return False

def send_email(otp: str, recipient_email: str, purpose: str = "verification"):
    """Send OTP email using Resend with fallback to SMTP."""
    html_content = get_otp_email_template(otp, purpose=purpose)
    subject = "Reset Your Password | BudgetBuddy OTP" if purpose == "password_reset" else "Email Verification | BudgetBuddy OTP"

    # 1. Try Resend if configured
    if Settings.RESEND_API_KEY:
        try:
            resend.api_key = Settings.RESEND_API_KEY
            params: resend.Emails.SendParams = {
                "from": "bb@developerabhishek.me",
                "to": recipient_email,
                "subject": subject,
                "html": html_content,
            }
            resend.Emails.send(params)
            logger.info(f"Email successfully sent via Resend to {recipient_email}")
            return
        except Exception as e:
            logger.warning(f"Resend send failed: {e}. Attempting SMTP fallback...")

    # 2. Fallback to SMTP
    smtp_sent = send_smtp_email(recipient_email, subject, html_content)
    if not smtp_sent:
        logger.error(f"All email dispatch methods failed for {recipient_email}")

File: emailService.ts
import nodemailer from 'nodemailer';

let transporter: nodemailer.Transporter | null = null;

const isPlaceholder = (val?: string): boolean => {
  if (!val) return true;
  const lower = val.toLowerCase().trim();
  return (
    lower.includes('yourgmail@gmail.com') ||
    lower.includes('your_16_character_app_password') ||
    lower.includes('your_google_app_password') ||
    lower.includes('xxxxxxxxxxxxxxxx') ||
    lower === ''
  );
};

export const getTransporter = (): nodemailer.Transporter => {
  const gmailUser = (process.env.GMAIL_USER || process.env.SMTP_USER || '').trim();
  const rawPass = (process.env.GMAIL_APP_PASSWORD || process.env.SMTP_PASS || '');
  const gmailAppPass = rawPass.replace(/\s+/g, '');

  const isConfigured = !isPlaceholder(gmailUser) && !isPlaceholder(gmailAppPass);

  if (isConfigured && gmailUser && gmailAppPass) {
    return nodemailer.createTransport({
      service: 'gmail',
      host: 'smtp.gmail.com',
      port: 465,
      secure: true,
      auth: {
        user: gmailUser,
        pass: gmailAppPass,
      },
    });
  }

  if (process.env.SMTP_HOST && process.env.SMTP_USER && !isPlaceholder(process.env.SMTP_USER)) {
    const port = parseInt(process.env.SMTP_PORT || '587', 10);
    return nodemailer.createTransport({
      host: process.env.SMTP_HOST,
      port,
      secure: port === 465,
      auth: {
        user: process.env.SMTP_USER,
        pass: process.env.SMTP_PASS || '',
      },
    });
  }

  return nodemailer.createTransport({
    jsonTransport: true,
  });
};

export const initEmailTransporter = () => {
  transporter = getTransporter();
  const gmailUser = (process.env.GMAIL_USER || process.env.SMTP_USER || '').trim();
  const rawPass = (process.env.GMAIL_APP_PASSWORD || process.env.SMTP_PASS || '');
  const isConfigured = !isPlaceholder(gmailUser) && !isPlaceholder(rawPass);

  console.log('[EmailService] GMAIL_USER configured:', isConfigured, isConfigured ? `(${gmailUser})` : '');
  console.log('[EmailService] GMAIL_APP_PASSWORD configured:', isConfigured);

  if (isConfigured) {
    transporter.verify((err) => {
      if (err) {
        console.warn('[EmailService] Gmail SMTP verification warning:', err.message);
      } else {
        console.log(`[EmailService] Gmail SMTP connection verified successfully for ${gmailUser}!`);
      }
    });
  }
};

export interface EmailParams {
  to: string;
  subject: string;
  title: string;
  message: string;
  detailsHtml?: string;
  actionUrl?: string;
  actionText?: string;
}

export const sendNotificationEmail = async (params: EmailParams): Promise<{ success: boolean; liveDispatched: boolean }> => {
  try {
    if (!transporter) {
      initEmailTransporter();
    }

    const { to, subject, title, message, detailsHtml, actionUrl, actionText } = params;

    const htmlContent = `
      <!DOCTYPE html>
      <html>
      <head>
        <meta charset="utf-8">
        <style>
          body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif; background-color: #0f172a; color: #f8fafc; margin: 0; padding: 20px; }
          .container { max-width: 580px; margin: 0 auto; background: #1e293b; border-radius: 12px; padding: 28px; border: 1px solid #334155; }
          .header { text-align: center; border-bottom: 1px solid #334155; padding-bottom: 16px; margin-bottom: 24px; }
          .brand { font-size: 24px; font-weight: bold; color: #3b82f6; text-decoration: none; }
          .title { font-size: 20px; font-weight: 700; color: #ffffff; margin-bottom: 12px; }
          .message { font-size: 15px; color: #cbd5e1; line-height: 1.6; margin-bottom: 20px; }
          .box { background: #0f172a; border-radius: 8px; padding: 16px; margin-bottom: 24px; border: 1px solid #334155; }
          .btn { display: inline-block; background-color: #2563eb; color: #ffffff !important; padding: 12px 24px; border-radius: 8px; font-weight: 600; text-decoration: none; text-align: center; }
          .footer { margin-top: 32px; text-align: center; font-size: 12px; color: #64748b; }
        </style>
      </head>
      <body>
        <div class="container">
          <div class="header">
            <span class="brand">BudgetBuddy</span>
          </div>
          <div class="title">${title}</div>
          <div class="message">${message}</div>
          ${detailsHtml ? `<div class="box">${detailsHtml}</div>` : ''}
          ${
            actionUrl
              ? `<div style="text-align: center; margin-top: 24px;">
                  <a href="${actionUrl}" class="btn">${actionText || 'View Dashboard'}</a>
                </div>`
              : ''
          }
          <div class="footer">
            <p>This is an automated security & financial notification from BudgetBuddy.</p>
            <p>&copy; ${new Date().getFullYear()} BudgetBuddy Personal Finance Manager</p>
          </div>
        </div>
      </body>
      </html>
    `;

    const fromAddress = process.env.GMAIL_USER && !isPlaceholder(process.env.GMAIL_USER)
      ? process.env.GMAIL_USER
      : 'alerts@budgetbuddy.app';

    const info = await transporter!.sendMail({
      from: `"BudgetBuddy" <${fromAddress}>`,
      to,
      subject,
      html: htmlContent,
    });

    console.log(`[EmailService] Notification email dispatched to ${to}: ${info?.messageId || 'Success'}`);
    return { success: true, liveDispatched: !isPlaceholder(process.env.GMAIL_USER) };
  } catch (error: any) {
    console.warn('[EmailService] Email notification fallback log:', error.message || error);
    return { success: true, liveDispatched: false };
  }
};

export const sendVerificationCodeEmail = async (params: {
  to: string;
  name: string;
  code: string;
}): Promise<{ success: boolean; liveDispatched: boolean }> => {
  const { to, name, code } = params;

  console.log('\n======================================================================');
  console.log('📧 [BUDGETBUDDY VERIFICATION CODE DISPATCH]');
  console.log(`To:   ${to.trim()}`);
  console.log(`Name: ${name || 'User'}`);
  console.log(`CODE: >>> ${code} <<<`);
  console.log('Expires in: 10 minutes');
  console.log('======================================================================\n');

  try {
    const activeTransporter = getTransporter();
    const gmailUser = (process.env.GMAIL_USER || process.env.SMTP_USER || '').trim();
    const isConfigured = !isPlaceholder(gmailUser) && !isPlaceholder(process.env.GMAIL_APP_PASSWORD);

    const subject = `Your BudgetBuddy Verification Code: ${code}`;
    const textContent = `Hello ${name || 'User'},\n\nYour BudgetBuddy verification code is: ${code}\n\nEnter this 6-digit code on the BudgetBuddy verification page to complete your sign-in.\n\nThis code will expire in 10 minutes.\n\nIf you did not request this verification code, please ignore this email.\n\nRegards,\nBudgetBuddy Team`;

    const htmlContent = `
      <!DOCTYPE html>
      <html>
      <head>
        <meta charset="utf-8">
        <meta name="viewport" content="width=device-width, initial-scale=1.0">
        <title>BudgetBuddy Verification Code</title>
      </head>
      <body style="margin: 0; padding: 20px; font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Helvetica, Arial, sans-serif; background-color: #f8fafc; color: #0f172a;">
        <table width="100%" border="0" cellspacing="0" cellpadding="0" style="max-width: 520px; margin: 0 auto; background-color: #ffffff; border: 1px solid #e2e8f0; border-radius: 12px; overflow: hidden; box-shadow: 0 4px 6px -1px rgba(0,0,0,0.05);">
          <tr>
            <td style="padding: 28px 32px; background-color: #0f172a; text-align: center;">
              <span style="font-size: 24px; font-weight: 800; color: #3b82f6; letter-spacing: -0.5px;">BudgetBuddy</span>
            </td>
          </tr>
          <tr>
            <td style="padding: 32px;">
              <h2 style="margin: 0 0 16px; font-size: 20px; font-weight: 700; color: #0f172a;">Hello ${name || 'User'},</h2>
              <p style="margin: 0 0 20px; font-size: 15px; line-height: 1.6; color: #334155;">
                Your 6-digit BudgetBuddy verification code is:
              </p>
              <div style="background-color: #eff6ff; border: 2px solid #3b82f6; border-radius: 10px; padding: 20px; text-align: center; margin: 24px 0;">
                <span style="font-family: 'SFMono-Regular', Consolas, Menlo, Monaco, monospace; font-size: 36px; font-weight: 800; letter-spacing: 8px; color: #1d4ed8;">${code}</span>
                <div style="font-size: 12px; color: #64748b; margin-top: 8px;">Valid for 10 minutes &bull; One-time use only</div>
              </div>
              <p style="margin: 0 0 12px; font-size: 14px; line-height: 1.6; color: #64748b;">
                Enter this code on the verification screen to complete your sign-in.
              </p>
              <p style="margin: 0; font-size: 13px; color: #94a3b8;">
                If you did not request this verification code, no action is needed.
              </p>
            </td>
          </tr>
          <tr>
            <td style="padding: 20px 32px; background-color: #f8fafc; border-top: 1px solid #f1f5f9; text-align: center; font-size: 12px; color: #64748b;">
              <p style="margin: 0 0 4px;">BudgetBuddy Personal Finance &bull; Security Notification</p>
              <p style="margin: 0;">&copy; ${new Date().getFullYear()} BudgetBuddy</p>
            </td>
          </tr>
        </table>
      </body>
      </html>
    `;

    const fromAddress = isConfigured && gmailUser ? gmailUser : 'security@budgetbuddy.app';

    if (isConfigured) {
      const info = await activeTransporter.sendMail({
        from: `"BudgetBuddy Security" <${fromAddress}>`,
        replyTo: fromAddress,
        to: to.trim(),
        subject,
        text: textContent,
        html: htmlContent,
        priority: 'high',
        headers: {
          'X-Priority': '1',
          'X-MSMail-Priority': 'High',
          Importance: 'High',
        },
      });

      console.log('[EmailService] Live verification email sent to:', to.trim(), 'MessageID:', info?.messageId);
      return { success: true, liveDispatched: true };
    } else {
      console.log('[EmailService] Demo/Simulation mode: code generated and logged above.');
      return { success: true, liveDispatched: false };
    }
  } catch (error: any) {
    console.warn('[EmailService] Gmail SMTP delivery failed:', error.message || error);
    return { success: false, liveDispatched: false };
  }
};
